src/module/CandidateSplitting.cpp

```diff
@@ -0,0 +1,154 @@
+#include "crpropa/module/CandidateSplitting.h"
+
+namespace crpropa {
+
+CandidateSplitting::CandidateSplitting() {
+	// no particle splitting if EnergyBins and NSplit are not specified
+	setNsplit(0);
+	setMinimalWeight(1.);
+}
+
+CandidateSplitting::CandidateSplitting(int n_split, double Emin, double Emax, double n_bins, double minWeight) {
+	// linear energy bins:
+	setNsplit(n_split);
+	setEnergyBins(Emin, Emax, n_bins, false);
+	setMinimalWeight(minWeight);
+}
+
+CandidateSplitting::CandidateSplitting(int n_split, double Emin, double Emax,  double n_bins, double minWeight, bool log) {
+	setNsplit(n_split);
+	setEnergyBins(Emin, Emax, n_bins, log);
+	setMinimalWeight(minWeight);
+}
+
+CandidateSplitting::CandidateSplitting(double SpectralIndex, double Emin, int nBins)  {
+	// to use with Diffusive Shock Acceleration
+
+	if (SpectralIndex <= 0){
+		throw std::runtime_error(
+				"CandidateSplitting: spectralIndex <= 0 !");
+	}
+
+	setNsplit(2); // always split in 2, calculate bins in energy for given spectrum:
+	double dE = pow(1./2, 1./(-SpectralIndex + 1)); 
+	setEnergyBinsDSA(Emin, dE, nBins);
+	setMinimalWeight(1./pow(2, nBins));
+
+}
+
+
+void CandidateSplitting::process(Candidate *c) const {
+
+	double currE = c->current.getEnergy(); 
+	double prevE = c->previous.getEnergy();
+
+	if (c->getWeight() <= minWeight){
+		// minimal weight reached, no splitting
+		return;
+	}
+	if (currE < Ebins[0] || n_split == 0 ){
+		// current energy is smaller than first bin -> no splitting
+		// or, number of splits = 0
+		return;
+	}
+	for (size_t i = 0; i < Ebins.size(); ++i){
+		
+		if( prevE < Ebins[i] ){
+			// previous energy is in energy bin [i-1, i]
+			if(currE < Ebins[i]){
+				//assuming that dE greater than 0, prevE and E in same energy bin -> no splitting
+				return;
+			}
+			// current energy is in energy bin [i,i+1] or higher -> particle splitting for each crossing
+			for (size_t j = i; j < Ebins.size(); ++j ){
+
+				// adapted from Acceleration Module:
+				c->updateWeight(1. / n_split); // * 1/n_split
+
+				for (int i = 1; i < n_split; i++) {
+				
+					ref_ptr<Candidate> new_candidate = c->clone(false);
+					new_candidate->parent = c;
+					uint64_t snr = Candidate::getNextSerialNumber();
+					new_candidate->setSerialNumber(snr);
+					new_candidate->previous.setEnergy(currE); // so that new candidate is not split again in next step!
+					//InteractionTag is PRIM, physically no new particles are created
+					c->addSecondary(new_candidate);
+					Candidate::setNextSerialNumber(snr + 1);
```

Comment:
Is this save for parallelization? I would expect a `#pragma omp critical`. Maybe the creation of [new candidates](https://github.com/CRPropa/CRPropa3/blob/cab3867959008514f6796681bca1fc1c6397d6c1/src/Candidate.cpp#L17) should be compared.

Reply:
And we should remove that from `Acceleration` as well :-)